scipy/optimize/_differentialevolution.py

```diff
@@ -689,9 +689,8 @@ def _ensure_constraint(self, trial):
         """
         make sure the parameters lie between the limits
         """
-        for index, param in enumerate(trial):
-            if param > 1 or param < 0:
-                trial[index] = self.random_number_generator.rand()
+        for index in np.where((trial < 0) | (trial > 1)):
```

Comment:
The improvement you suggest has a few issues with it. The main one is that for all the entries that are outside the range, it uses the same random number as the replacement.
Use
```
trial = np.where((trial < 0) | (trial > 1), np.random.rand(trial.size), trial)
```
instead.